components/masking_functions/src/masking_functions/registration_routines.cpp

```diff
@@ -0,0 +1,1156 @@
+      const masking_functions::charset_string &masking_char) const override {
+    auto sresult =
+        masking_functions::mask_inner(cs_str, 0, 36 - 8, masking_char);
+    sresult =
+        masking_functions::mask_inner(sresult, 9, 36 - 9 - 4, masking_char);
+    sresult = masking_functions::mask_inner(sresult, 9 + 5, 36 - 9 - 5 - 4,
+                                            masking_char);
+    sresult = masking_functions::mask_inner(sresult, 9 + 5 + 5,
+                                            36 - 9 - 5 - 5 - 4, masking_char);
+    sresult =
+        masking_functions::mask_inner(sresult, 9 + 5 + 5 + 5, 0, masking_char);
+
+    return sresult;
+  }
+};
+
+//
+// gen_blocklist(string, string, string)
+//
+class gen_blocklist_impl {
+ public:
+  gen_blocklist_impl(mysqlpp::udf_context &ctx) {
+    if (ctx.get_number_of_args() != 3)
+      throw std::invalid_argument{
+          "Wrong argument list: gen_blocklist(string, string, string)"};
+
+    ctx.mark_result_nullable(true);
+    ctx.mark_result_const(false);
+
+    // arg1 - dictionary
+    ctx.mark_arg_nullable(0, false);
+    ctx.set_arg_type(0, STRING_RESULT);
+
+    ctx.mark_arg_nullable(1, false);
+    ctx.set_arg_type(1, STRING_RESULT);
+
+    ctx.mark_arg_nullable(2, false);
+    ctx.set_arg_type(2, STRING_RESULT);
+
+    mysqlpp::udf_context_charset_extension charset_ext{
+        mysql_service_mysql_udf_metadata};
+    set_return_value_collation_from_arg(charset_ext, ctx, 0);
+  }
+
+  mysqlpp::udf_result_t<STRING_RESULT> calculate(
+      const mysqlpp::udf_context &ctx) {
+    const auto cs_term = extract_charset_string_from_arg(ctx, 0);
+    const auto cs_dict_a = extract_charset_string_from_arg(ctx, 1);
+    const auto cs_dict_b = extract_charset_string_from_arg(ctx, 2);
+
+    {
+      masking_functions::sql_context sql_ctx{
+          global_command_services::instance()};
+
+      auto query =
+          global_query_builder::instance()
+              .select_random_term_for_dictionary_and_term(cs_dict_a, cs_term);
+      auto sresult = sql_ctx.query_single_value(query);
+
+      if (!sresult) {
+        return {std::string{cs_term.get_buffer()}};
+      }
+    }
+
+    masking_functions::sql_context sql_ctx{global_command_services::instance()};
+
+    auto query =
+        global_query_builder::instance().select_random_term_for_dictionary(
+            cs_dict_b);
+    auto sresult = sql_ctx.query_single_value(query);
+
+    if (sresult && sresult->size() > 0) {
+      masking_functions::charset_string utf8_result{
+          global_string_services::instance(), *sresult,
+          masking_functions::charset_string::utf8mb4_collation_name};
+      masking_functions::charset_string conversion_buffer;
+      const auto &cs_result = masking_functions::smart_convert_to_collation(
+          utf8_result, cs_term.get_collation(), conversion_buffer);
+      return {std::string{cs_result.get_buffer()}};
+    } else {
+      return std::nullopt;
+    }
+  }
+};
```

Comment:
The only test coverage I added for this is to check how calls to dictionary manipulation UDFs are replicated in different modes (SBR/RBR/MIXED).
See, `component_masking_functions.rpl_function_determinism_xxx` for more info.
Still nothing regarding Stored Procedures though. Probably is worth adding.

Reply:
Let me elaborate a bit. My primarily concern here is what will happen to binary log and MySQL Server in general when one of those functions will update table in such out-of-band fashion in the middle of another statement execution. Since access to those tables are not visible to prelocking algorithm there is a chance that there will be some kind of corruption in this case (if context of the same connection is used for this out-of-band access).

Reply:
Per Slack discussion this can be addressed as a follow-up patch. Manual testing confirms that everything seems to be working ok.